[DirectX] Set DXIL Version using shader model version in compilation target profile

bharadwajy opened the pull request:
An earlier commit provided a way to decouple DXIL version from Shader Model 
version by representing the DXIL version as `SubArch` in the DXIL Target Triple 
and adding corresponding DXIL `Arch` types as valid. 

This change constructs the DXIL target triple with DXIL version that is deduced 
from the Shader Model version specified in the compilation target profile in a dxc 
driver invocation flow.

Updates to tests necessitated by the above change are also included:
* Arch string `dxil` in all HLSL and DirectX tests is updated to include appropriate
    version number string.
* Normalization tests of triple strings with DXIL Arch that now has version number 
  are separated in `TripleTest.cpp`.

PR #90809 adds the functionality that does not require explicit specification of DXIL version in
a target triple specified on command-line.

Commit:
Set DXIL Version based on shader model provided in compilation target profile specified

Update dxil arch string in all HLSL and DirectX tests

Separate out normalization tests of triple strings with DXIL Arch
that now has version number.

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -98,9 +100,47 @@ std::optional<std::string> tryParseProfile(StringRef Profile) {
+    break;
+  }
+  if (SubArch == llvm::Triple::NoSubArch) {
+    report_fatal_error(
```
The frontend should not call `report_fatal_error`. If this error is possible we should emit a driver diagnostic (see the `D.Diag` calls elsewhere in this file).

If this can't happen, we should instead assert.

If this error can occur we need a test that verifies the error condition.

bharadwajy replied:
> The frontend should not call `report_fatal_error`. If this error is possible we should emit a driver diagnostic (see the `D.Diag` calls elsewhere in this file).
> 
> If this can't happen, we should instead assert.
> 
> If this error can occur we need a test that verifies the error condition.

Pushed a change to address the feedback. Thanks!

tex3d commented on clang/unittests/Driver/DXCModeTest.cpp:
```diff
@@ -68,25 +68,25 @@ TEST(DxcModeTest, TargetProfileValidation) {
+  validateTargetProfile("-Tas_6_7", "dxilv1.7--shadermodel6.7-amplification",
                         InMemoryFileSystem, Diags);
-  validateTargetProfile("-Tlib_6_x", "dxil--shadermodel6.15-library",
+  validateTargetProfile("-Tlib_6_8", "dxilv1.8--shadermodel6.8-library",
```
There is, in fact, a lib_6_x target that maps to a more unconstrained offline linking target (with minor version = 15) which is not expected to be compatible between compiler versions.  Do we care to model that in clang at this time?  This change removes this target, so I just wanted to know if that was a deliberate decision.

bharadwajy replied:
> There is, in fact, a lib_6_x target that maps to a more unconstrained offline linking target (with minor version = 15) which is not expected to be compatible between compiler versions. Do we care to model that in clang at this time? This change removes this target, so I just wanted to know if that was a deliberate decision.

What is the expected DXIL version of DXIL Ops generated using the target profile `lib_6_x`?

llvm-beanz replied:
`lib_6_x` should basically map to "whatever the latest DXIL version is"

bharadwajy replied:
> `lib_6_x` should basically map to "whatever the latest DXIL version is"

In that case should the call be changed to 
```
validateTargetProfile("-Tlib_6_x", "dxilv1.8--shadermodel6.8-library",
                        InMemoryFileSystem, Diags);
```

and the option `-T<kind>6_x` be considered equivalent to `-T<kind>6_8` (as of now)?

llvm-beanz replied:
I think the change you proposed is correct (`-Tlib_6_x` maps to `dxilv1.8--shadermodel6.15-library`), but that is crucially _not_ making it equivalent to `-Tlib_6_8`.

Today when DXC compiles the `lib_6_8` target the resulting binary encodes the shader model as 6.8 and DXIL version as 1.8.

Today when DXC compiles the `lib_6_x` target the resulting binary encodes the shader model as 6.15 and DXIL version as 1.8.

Which means:
* `lib_6_8` -> `dxilv1.8--shadermodel6.8-library`
* `lib_6_x` -> `dxilv1.8--shadermodel6.15-library`

bharadwajy replied:
> `lib_6_x` should basically map to "whatever the latest DXIL version is"

Pushed a change to consider `lib_6_x` as a valid profile equivalent to `lib_6_8`.

Commit:
Incorporate PR feedback
 - Delete call to report_fatal_error and return std::nullopt. As a result
   appropriate unsupported SM version diagnostic is reported.
 - Add the requirement of SubArch to be set for a DXIL Triple to be deemed legal.
 - Add test to verify invalid profile error.

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -12,7 +12,10 @@
 #include "clang/Driver/DriverDiagnostic.h"
 #include "clang/Driver/Job.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/ErrorHandling.h"
```
This header should not be needed.

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -35,6 +38,11 @@ bool isLegalShaderModel(Triple &T) {
     return false;
 
+  // DXIL Version should be set as SubArch
+  if (T.getSubArch() == Triple::NoSubArch) {
+    return false;
+  }
```
nit:
```suggestion
  if (T.getSubArch() == Triple::NoSubArch)
    return false;
```

see: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -12,7 +12,10 @@
+#include "llvm/Support/ErrorHandling.h"
 #include "llvm/TargetParser/Triple.h"
+#include <optional>
+#include <string>
```
Do we need <string>?

llvm-beanz commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -98,9 +106,46 @@ std::optional<std::string> tryParseProfile(StringRef Profile) {
+    SubArch = llvm::Triple::DXILSubArch_v1_8;
+    break;
+  }
+  if (SubArch == llvm::Triple::NoSubArch) {
```
Should we fold this into a `default` case for the switch statement?

Commit:
Incorporate additional PR feedback.

Commit:
Restore support for target profile lib_6_x

llvm-beanz commented on clang/unittests/Driver/DXCModeTest.cpp:
```diff
@@ -68,25 +68,25 @@ TEST(DxcModeTest, TargetProfileValidation) {
+  validateTargetProfile("-Tas_6_7", "dxilv1.7--shadermodel6.7-amplification",
                         InMemoryFileSystem, Diags);
-  validateTargetProfile("-Tlib_6_x", "dxil--shadermodel6.15-library",
+  validateTargetProfile("-Tlib_6_x", "dxilv1.8--shadermodel6.15-library",
```
Seems like we should have a test case for `lib_6_8` too since they're not the same.

bharadwajy replied:
> Seems like we should have a test case for `lib_6_8` too since they're not the same.

Added a test case.

Commit:
Add test case for lib_6_8 in DXC driver tests

bogner commented on clang/test/Driver/hlsl-lang-targets.hlsl:
```diff
@@ -2,39 +2,45 @@
+// RUN: not %clang -target dxilv1.2--shadermodel6.2-msvc %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-BAD-ENV %s
+
+// Invaliddxil target with no version number
+// RUN: not %clang -target dxil--shadermodel6.2-msvc %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-BAD-TARGET %s
```
So is the `dxil` arch with no subarch just completely disallowed then? This is unusual compared to other targets, and also a bit inconvenient for testing. If I'm trying something locally and switching between SM6.2 and SM6.0 I now need to update two parts of the triple on my command line even though these shader models and DXIL versions have  a 1-1 mapping.

I also worry that this error message will be pretty confusing. IIUC it will print "HLSL code generation is unsupported for target 'dxil--shadermodel6.2-msvc'". We should arguably have an error message specifically saying that the DXIL version is invalid for the shader model. This might be covered by my other comment.

bharadwajy replied:
> So is the `dxil` arch with no subarch just completely disallowed then? 

Yes.

> This is unusual compared to other targets, and also a bit inconvenient for testing. If I'm trying something locally and switching between SM6.2 and SM6.0 I now need to update two parts of the triple on my command line even though these shader models and DXIL versions have a 1-1 mapping.

The intent is to avoid implicit mapping of DXIL and SM versions. Allowing `dxil` to be valid would require an implicit assumption of the DXIL version for such a specification - and can be cause for confusion. Specifying the targeted DXIL and SM versions **explicitly** - particularly since they have been decoupled, seems appropriate.

python3kgae replied:
I'll vote for allowing implicit mapping since that is the common case.
Even though they're decoupled, allowing a default value would match what other targets do.
Like spirv, it is ok to not say spirv32 for their test instead of require spirv32_v1.0.

bogner commented on clang/test/Driver/hlsl-lang-targets.hlsl:
```diff
@@ -2,39 +2,45 @@
-// RUN: not %clang -target dxil--shadermodel6.2 %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-NO-ENV %s
+// RUN: not %clang -target dxilv1.2-shadermodel6.2 %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-NO-ENV %s
+// RUN: not %clang -target dxilv1.2--shadermodel6.2 %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-NO-ENV %s
+// RUN: not %clang -target dxilv1.2--shadermodel6.2 %s -S -o /dev/null 2>&1 | FileCheck --check-prefix=CHECK-NO-ENV %s
```
What should happen if someone specifies `dxilv1.0--shadermodel6.7` and then uses some SM6.7 feature that isn't representable in dxil 1.0? We should probably validate that the dxil version and shadermodel are compatible.

A 1-1 mapping for validating the shader model vs the dxil version is probably sufficient / matches DXC. It's also possible we could do better, but I don't know that it's worth the effort of cataloguing compatibility for that.

bharadwajy replied:
> What should happen if someone specifies `dxilv1.0--shadermodel6.7` and then uses some SM6.7 feature that isn't representable in dxil 1.0? We should probably validate that the dxil version and shadermodel are compatible.

I'd expect such validation to be done during analysis and/or codegen passes - potentially specified in `DXIL.td` where the properties of DXIL Operations are specified.
 
> A 1-1 mapping for validating the shader model vs the dxil version is probably sufficient / matches DXC. It's also possible we could do better, but I don't know that it's worth the effort of cataloguing compatibility for that.

Such cataloguing would be part of `DXIL.td`.

bogner commented on llvm/lib/TargetParser/Triple.cpp:
```diff
@@ -115,6 +115,30 @@ StringRef Triple::getArchName(ArchType Kind, SubArchType SubArch) {
+      return "dxilv1.8";
+    default:
+      report_fatal_error("Unknown DXIL Version",
+                         /*gen_crash_diag=*/false);
```
I don't think we should crash if we have an architecture without a subarch. We don't for any of the other architectures.

bogner commented on llvm/unittests/TargetParser/TripleTest.cpp:
```diff
@@ -1171,18 +1181,27 @@ TEST(TripleTest, Normalization) {
+    // Exclude testing DXIL Arch since arch type name for Triple::dxil depends
+    // on SubArch and is no longer "dxil". Instead, normalization tests for
+    // Triple::dxil are added at an earlier location in this test function.
+    if (Arch != Triple::dxil) {
```
This is more evidence that treating DXIL specially here is problematic.

bogner commented on clang/lib/Driver/ToolChains/HLSL.cpp:
```diff
@@ -98,9 +103,47 @@ std::optional<std::string> tryParseProfile(StringRef Profile) {
+  case 8:
+  case OfflineLibMinor:
+    // Always consider minor version x as the latest supported minor version
+    SubArch = llvm::Triple::DXILSubArch_v1_8;
```
It might make updates here less error prone if we added `DXILSubArch_v1_x = DXILSubArch_v1_8` or `LastDXILSubArch = DXILSubArch_v1_8` to the enum in Triple.h, depending on which of those makes more logical sense.

bharadwajy replied:
> It might make updates here less error prone if we added `DXILSubArch_v1_x = DXILSubArch_v1_8` or `LastDXILSubArch = DXILSubArch_v1_8` to the enum in Triple.h, depending on which of those makes more logical sense.

OK.

bogner reviewed (changes requested):
There's a lot of churn here so I'm not 100% sure I saw all of the real changes, but I think I got them all. It might be nice to move the NFC parts into their own commit (if we actually need them, see my other comments).

I don't think we should treat the DXIL architecture differently than all of the other architectures in LLVM like this. More on that in the comments.

damyanp commented:
Now the #90809 has landed, is this PR still relevant?

bharadwajy commented:
> Now the #90809 has landed, is this PR still relevant?

No. I was just about to close it. Thanks!

Closing as an expanded implementation has been merged.
